### DISCUSSION TITLE
Binding to sorted IObservableList<T>

### DISCUSSION THREAD

**Discussion by Jeremy-Code-F** [AvaloniaUI/Avalonia#9109] [Q&A] [answered]
Hi,

I'm trying to implement sorting for a TreeView. I've got a simple MRE of the problem I'm seeing [here](https://bitbucket.org/jeremycodef/avalonia-sorting/src/master/)

I create and initialize my ```ObservableCollection``` in my ```MainWindowViewModel``` constructor

```
 public class MainWindowViewModel : ViewModelBase
    {
        public string Greeting => "Welcome to Avalonia!";

        public ObservableCollection<TestViewModel> CollectionToSort { get; set; } =
            new ObservableCollection<TestViewModel>();
        public IObservableList<TestViewModel> SortedList { get; set; }

        public MainWindowViewModel()
        {
            this.CollectionToSort.Add(new TestViewModel
            {
                Id = Guid.NewGuid(),
                OrderIndex = 2
            });
            this.CollectionToSort.Add(new TestViewModel
            {
                Id = Guid.NewGuid(),
                OrderIndex = 0
            });
            this.CollectionToSort.Add(new TestViewModel
            {
                Id = Guid.NewGuid(),
                OrderIndex = 1
            });
...
```
Based on a conversation on [gitter](https://gitter.im/AvaloniaUI/Avalonia?at=5e1f300132c2e05ecc4ae34b) while trying to figure out how to accomplish the sorting it looks like I need to use [DynamicData](https://gitter.im/AvaloniaUI/Avalonia?at=5e1f530ddf9e137d58ec5bb3)

So I setup an ```IObservableList<TestViewModel>``` and initialize it at the end of ```MainWindowViewModel```

```
SortedList = CollectionToSort.ToObservableChangeSet()
                .Sort(SortExpressionComparer<TestViewModel>.Ascending(t => t.OrderIndex))
                .AsObservableList();
            Console.WriteLine();
```
 

But it doesn't look like the binding for an ```IObservableList``` works the same way that it does for ```ObservableCollection```

```
<TreeView Items="{Binding CollectionToSort}">
       <TreeView.DataTemplates>
           <DataTemplate DataType="vm:TestViewModel">
               <TextBlock Text="{Binding DisplayText}"></TextBlock>
           </DataTemplate>
       </TreeView.DataTemplates> 
    </TreeView>
```

ObservableCollection - 
![image](https://user-images.githubusercontent.com/45114285/194165682-c9759364-ac03-4947-8ace-3a4a1bea4d36.png)

IObservableList - 
![image](https://user-images.githubusercontent.com/45114285/194165751-b680e08d-80f5-465d-90b5-36b812c8de70.png)

Am I missing something in this setup? Or is this even the right way to approach sorting?

Thanks!

**Comment by timunie** [comment, marked as answer]
Let me suggest three things:

1. You can remove the setter from your collections or make them private. You don't want them to be changed at runtime. 

2. DynamicData works a bit different. I know a sample for Xamarin, which can be used in Avalonia almost 1:1, see https://www.xamboy.com/2021/01/25/filtering-and-sorting-with-dynamic-data-in-xamarin-forms-part-2/

3. If you expect a huge amount of data, consider to use TreeDataGrid as the normal TreeView is not vitualized, which means it can get slow with a lot of items. 

Happy coding
Tim

**Reply by Jeremy-Code-F** [reply]
Tyvm! Using that resource I was able to get something working, I also went ahead and added a pr with docs for a simple example of sorting/filtering for future people.
https://docs.avaloniaui.net/guides/deep-dives/reactiveui/binding-to-sorted-filtered-list
